In [9]:
import requests
from bs4 import BeautifulSoup
from lxml import html
import pandas as pd
import nltk
from string import punctuation
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
stop_words = set( stopwords.words('english'))
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import spacy 
import numpy as np
import string
import vaderSentiment

### Cleaning the data

In [12]:
reviews_df = pd.read_csv('reviews.csv')
reviews_df[:10]

product_name  \
0  Kentucky Brunch Brand Stout   
1  Kentucky Brunch Brand Stout   
2  Kentucky Brunch Brand Stout   
3  Kentucky Brunch Brand Stout   
4  Kentucky Brunch Brand Stout   
5  Kentucky Brunch Brand Stout   
6  Kentucky Brunch Brand Stout   
7  Kentucky Brunch Brand Stout   
8  Kentucky Brunch Brand Stout   
9  Kentucky Brunch Brand Stout   

                                      product_review  user_rating  
0  2016 Silver Wax. Aroma has whiskey, maple, tof...         4.80  
1  The beer pours Pitch Black with a frothy tan h...         4.74  
2  Probably the smoothest beer I have ever had. S...         4.68  
3  Dark black, very thick, a little bit of tan he...         5.00  
4  Poured black as ink with thin ruby edges at 58...         4.97  
5  she’s got heat....but man is she somethin nice...         4.75  
6  Amazing brew. The maple aroma pours out of thi...         5.00  
7  Finally got to try the white whale. Pours like...         4.30  
8  Had this on tap at the 12-15-18 release in Dec...         4.85  
9  Had this on tap At the KBBS release 12/15/18 p...         4.89

In [14]:
#Checking if there are 25 reviews of each product
reviews_df.groupby(['product_name']).count()

product_review  \
product_name                                                         
3rd Anniversary Imperial IPA                                    25   
4th Anniversary                                                 22   
A Deal With The Devil                                           25   
Aaron                                                           25   
Abner                                                           25   
Abrasive Ale                                                    25   
Abraxas                                                         25   
Abricot Du Fermier                                              25   
Abt 12                                                          25   
Adam From The Wood                                              25   
All That Is And All That Ever Will Be                           25   
Alter Ego                                                       25   
Anabasis                                                        20   
Ann                                                             25   
Apple Brandy Barrel Noir                                        25   
Art                                                             25   
Assassin                                                        25   
Atrial Rubicite                                                 25   
Aurelian Lure                                                   25   
Axe Man                                                         25   
BBADL (Bourbon Barrel Aged Dark Lord Imperial S...              25   
BDCS                                                            25   
Bad Boy                                                         25   
Barrel Aged Bomb!                                               25   
Barrel Aged Imperial German Chocolate Cupcake S...              25   
Barrel-Aged Abraxas                                             25   
Barrel-Aged Silhouette                                          25   
Barrel-Aged Sump Coffee Stout                                   25   
Beatification                                                   25   
Beer Geek Vanilla Shake - Bourbon Barrel-Aged                   25   
...                                                            ...   
Sunday Brunch                                                   25   
Supplication                                                    25   
Susan                                                           25   
Swish                                                           25   
Temptation                                                      25   
Ten FIDY - Bourbon Barrel Aged                                  25   
The Abyss                                                       25   
The Broken Truck                                                25   
The Peach                                                       25   
The Rusty Nail                                                  25   
Thicket                                                         25   
Trappist Westvleteren 12 (XII)                                  25   
Trappist Westvleteren 8 (VIII)                                  25   
Trappistes Rochefort 10                                         25   
Triple Shot                                                     25   
Triple Sunshine                                                 25   
Tweak                                                           25   
Upper Case                                                      25   
V.S.O.J.                                                        25   
Very GGGreennn                                                  23   
Very Green                                                      25   
Very Hazy                                                       25   
Vicinity                                                        25   
Weihenstephaner Hefeweissbier                                   25   
West Ashley                                                   

In [15]:
reviews_df.columns

Index(['product_name', 'product_review', 'user_rating'], dtype='object')

In [18]:
reviews_df = reviews_df.dropna()
reviews_df['cleaned_review'] = reviews_df['product_review'].apply(lambda x :x.translate(str.maketrans('', '', string.punctuation)))
reviews_df['cleaned_review'] = reviews_df['cleaned_review'].apply(lambda x :x.lower())

reviews_df['cleaned_review'] = reviews_df['cleaned_review'].apply(word_tokenize).apply(set).apply(list)

def remove_stopwords(s):
    return [w for w in s if not w in stop_words] 
    
reviews_df['cleaned_review'] =  reviews_df['cleaned_review'].apply(remove_stopwords)

### Word Frequency Analysis and Attribute Selection

In [19]:
#Creating a combined list of all words
all_words = []
for i in range(len(reviews_df)):
    all_words+=reviews_df['cleaned_review'][i]
from nltk import FreqDist
word_freq = nltk.FreqDist(all_words)

In [20]:
word_freq.most_common()

[('head', 3407),
 ('beer', 2785),
 ('taste', 2595),
 ('pours', 1725),
 ('nose', 1666),
 ('sweet', 1661),
 ('dark', 1621),
 ('like', 1598),
 ('one', 1596),
 ('carbonation', 1586),
 ('finish', 1547),
 ('chocolate', 1536),
 ('aroma', 1517),
 ('nice', 1445),
 ('mouthfeel', 1444),
 ('good', 1430),
 ('well', 1418),
 ('body', 1396),
 ('overall', 1374),
 ('light', 1355),
 ('lacing', 1352),
 ('bottle', 1349),
 ('medium', 1314),
 ('notes', 1310),
 ('black', 1290),
 ('flavor', 1285),
 ('smooth', 1235),
 ('white', 1224),
 ('vanilla', 1184),
 ('bit', 1167),
 ('color', 1166),
 ('fruit', 1148),
 ('bourbon', 1145),
 ('little', 1141),
 ('glass', 1136),
 ('great', 1130),
 ('orange', 1119),
 ('flavors', 1097),
 ('really', 1097),
 ('thick', 1096),
 ('feel', 1093),
 ('poured', 1077),
 ('smell', 1069),
 ('coffee', 1043),
 ('citrus', 1007),
 ('bitterness', 944),
 ('barrel', 932),
 ('oak', 903),
 ('brown', 892),
 ('creamy', 874),
 ('much', 870),
 ('sweetness', 856),
 ('hazy', 825),
 ('malt', 818),
 ('dry', 78

Selected attributes: 
1. **Crisp** - Highly carbonated; effervescent
2. **Hoppy** - Herbal, earthy, spicy, or citric aromas and flavors of hops
3. **Robust** - Rich and full-bodied


### Similarity analysis with the 3-attribute set and the reviews.

In [33]:
import spacy
import en_core_web_lg
nlp = en_core_web_lg.load()

In [34]:
def join_words(comment):   
    """Joins the tokenized words to a sentence"""
    return " ".join(comment) 

reviews_df['joined_review'] = reviews_df['cleaned_review'].map(join_words)

In [35]:
#Calculate similarity with pre-processing functions
def calculate_similarity(comment):
    """Compute similarity score"""
    base = nlp(comment)
    compare = nlp(input_attributes)
    return base.similarity(compare)

In [36]:
input_list = ['crisp', 'hoppy', 'robust']
input_attributes =  " ".join(input_list)
reviews_df['similarity'] = reviews_df['joined_review'].map(calculate_similarity)

In [37]:
reviews_df.head(10)

product_name  \
0  Kentucky Brunch Brand Stout   
1  Kentucky Brunch Brand Stout   
2  Kentucky Brunch Brand Stout   
3  Kentucky Brunch Brand Stout   
4  Kentucky Brunch Brand Stout   
5  Kentucky Brunch Brand Stout   
6  Kentucky Brunch Brand Stout   
7  Kentucky Brunch Brand Stout   
8  Kentucky Brunch Brand Stout   
9  Kentucky Brunch Brand Stout   

                                      product_review  user_rating  \
0  2016 Silver Wax. Aroma has whiskey, maple, tof...         4.80   
1  The beer pours Pitch Black with a frothy tan h...         4.74   
2  Probably the smoothest beer I have ever had. S...         4.68   
3  Dark black, very thick, a little bit of tan he...         5.00   
4  Poured black as ink with thin ruby edges at 58...         4.97   
5  she’s got heat....but man is she somethin nice...         4.75   
6  Amazing brew. The maple aroma pours out of thi...         5.00   
7  Finally got to try the white whale. Pours like...         4.30   
8  Had this on tap at the 12-15-18 release in Dec...         4.85   
9  Had this on tap At the KBBS release 12/15/18 p...         4.89   

                                      cleaned_review  \
0  [barrel, taste, umami, cacao, aroma, toffee, w...   
1  [creams, vanilla, frothy, pours, bottle, tan, ...   
2  [ever, alcohol, vanilla, taste, barrel, notes,...   
3  [ever, alcohol, either, blow, vanilla, sweet, ...   
4  [astringency, fizzy, strong, velvety, whole, b...   
5  [somethin, barrel, nose, booze, flavor, maple,...   
6  [brew, pours, bottle, aroma, brownish, af, bla...   
7  [breweries, pepper, like, think, pours, tg, gr...   
8  [barrel, release, high, maybe, 121518, good, m...   
9  [great, barrel, taste, release, silky, 121518,...   

                                       joined_review  similarity  
0  barrel taste umami cacao aroma toffee wood map...    0.635786  
1  creams vanilla frothy pours bottle tan feel bl...    0.602877  
2  ever alcohol vanilla taste barrel notes probab...    0.592445  
3  ever alcohol either blow vanilla sweet taste l...    0.536490  
4  astringency fizzy strong velvety whole booze 5...    0.638830  
5  somethin barrel nose booze flavor maple molass...    0.584269  
6  brew pours bottle aroma brownish af black carb...    0.626235  
7  breweries pepper like think pours tg green sme...    0.481698  
8  barrel release high maybe 121518 good mix arom...    0.589091  
9  great barrel taste release silky 121518 attemp...    0.630627

In [38]:
top300_reviews = reviews_df.sort_values(by='similarity', ascending=False)[0:300]

In [39]:
top300_reviews

product_name  \
2487                               Notorious Triple IPA   
2290                                         Flora Plum   
4107                                     Saison Bernice   
5609                                           Vicinity   
4330                                            Bad Boy   
1691                                       Doppelganger   
5784                                            Axe Man   
1044                                           King Sue   
1717                                            Ephraim   
1557                              Society & Solitude #4   
4524                                   Fuzzy Baby Ducks   
4422                                            Crusher   
3654                       Mosaic Dry Hopped Fort Point   
1225                                       Focal Banger   
2065              Galaxy Dry Hopped Fort Point Pale Ale   
1118                                      Double Galaxy   
4096                                        Hopslam Ale   
5360                              Hop JuJu Imperial IPA   
5154             Double Dry Hopped All Citra Everything   
3013                                         Pseudo Sue   
1784              Double Dry Hopped Fort Point Pale Ale   
4416                                            Crusher   
1116                                      Double Galaxy   
5010  Somewhere, Something Incredible Is Waiting To ...   
5411                                Fort Point Pale Ale   
4622                                              Bodhi   
4931                               Ghost In The Machine   
4306                                Double Barrel Jesus   
3362                                  Permanent Funeral   
4120                                     Saison Bernice   
...                                                 ...   
5097                      Casey / Side Project - Leaner   
5401                                Fort Point Pale Ale   
3291                                    Mocha Wednesday   
1707                                       Doppelganger   
6115                                          Scaled Up   
1889                                         Keene Idea   
4824                                       Black & Wild   
5769                                            Axe Man   
1531                                               Haze   
939                                              Westly   
2483                               Notorious Triple IPA   
1918                                           Anabasis   
3373                                  Permanent Funeral   
1261                  Double Dry Hopped Congress Street   
1965                                        Triple Shot   
6035                             Double Orange Starfish   
844                                         Zombie Dust   
1184                                              Citra   
6113                                          Scaled Up   
5423                         Bourbon Barrel Black Maple   
2097                                   The Broken Truck   
1077                                           Headroom   
4123                                     Saison Bernice   
1262                  Double Dry Hopped Congress Street   
2156                                          Alter Ego   
5510    Imperial Eclipse Stout - Elijah Craig (12 Year)   
1877                                         Keene Idea   
6001                                         Mastermind   
841                                         Zombie Dust   
2444                               Beyond Good And Evil   

                                         product_review  user_rating  \
2487  Had on tap. Just a luscious and delicious dipa...         4.49   
2290  2018 release.  A: Pours a very light pale, haz...         4.44   
4107  Tasted from bottle. Classic farmhouse saison. ...         4.44   
5609  Pours a golden peach with a fluffy fast dissip...         4.31   
4330  Shared from a grow

### Sentiment analysis on top 300 reviews and Sorting them (high to low) by the sentiment scores.

In [40]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
analyser = SentimentIntensityAnalyzer()

In [41]:
def sentiment_analyzer_scores(sentence):
    """Generate seniment score"""
    scores = analyser.polarity_scores(sentence)
    return scores['compound']

top300_reviews['sentiment'] = top300_reviews['product_review'].map(sentiment_analyzer_scores)

In [43]:
top300_reviews_sentiment = top300_reviews.sort_values(by='sentiment', ascending=False)

In [44]:
top300_reviews_sentiment

product_name  \
3458                                 Darkness   
1626                       Oude Geuze Vintage   
4930                     Ghost In The Machine   
5044                          Emerald Grouper   
3291                          Mocha Wednesday   
4654                                      Art   
3512                   Montmorency Vs Balaton   
5315                                    Flora   
5827                                  Thicket   
2497                     Notorious Triple IPA   
1866                               Keene Idea   
1824                                   Clover   
1386                          Sip Of Sunshine   
2483                     Notorious Triple IPA   
5621  Jackie O's / Side Project - Appervation   
3258                             Abrasive Ale   
1877                               Keene Idea   
6023                   Double Orange Starfish   
608                                    Dinner   
4056                                Last Snow   
3216                               Gggreennn!   
4824                             Black & Wild   
6113                                Scaled Up   
5294                                    Flora   
1443                                    Swish   
5145   Double Dry Hopped All Citra Everything   
6047               Dorothy (Wine Barrel Aged)   
5353                    Hop JuJu Imperial IPA   
4630                   Barrel-Aged Silhouette   
5405                      Fort Point Pale Ale   
...                                       ...   
4524                         Fuzzy Baby Ducks   
3061                          Pugachev Royale   
4558                                    Sosus   
2918                            In Perpetuity   
4041                            Double Nelson   
828                               Zombie Dust   
1108                            Double Galaxy   
5401                      Fort Point Pale Ale   
4509                         Fuzzy Baby Ducks   
5609                                 Vicinity   
1261        Double Dry Hopped Congress Street   
1194                                    Citra   
5611                                 Vicinity   
1262        Double Dry Hopped Congress Street   
5097            Casey / Side Project - Leaner   
1462                                    Swish   
3342                            Scaled Way Up   
1225                             Focal Banger   
5791                                  Axe Man   
1291                                    Green   
104                              Heady Topper   
1557                    Society & Solitude #4   
1242                             Focal Banger   
5784                                  Axe Man   
965                             Double Citra®   
5391                            Second Fiddle   
1118                            Double Galaxy   
2923                            In Perpetuity   
2142                                Alter Ego   
5983                        Darkstar November   

                                         product_review  user_rating  \
3458  750 ml bottle into snifter, bottled on 9/25/20...         4.36   
1626  750ml bottle, 2008 vintage, poured into a Drie...         4.58   
4930  12 ounce bottle into tulip glass, bottled on 8...         4.28   
5044  16 ounce can into tulip glass, canned on 9/24/...         4.23   
3291  750 ml bottle into snifter, bottled on 1/27/20...         4.20   
4654  375ml bottle, Batch 5, dated March 18, 2015, p...         4.67   
3512  500 ml bottle into signature tulip glass; Blen...         4.37   
5315  750ml bottle, Batch #6, poured into a Hill Far...         4.51   
5827  375ml bottle, generously shared by Erik, poure...         4.19   
2497  had this on tap at the mule bar in Winooski VT...         4.34   
1866  16oz can, dated 08.05.19, poured into a 14oz s...         4.33   
1824  375ml bottle, poured into a Hill Farmstead ove...         4.56   
1386  Pours a mostly clear albeit cloudy orange colo...   

### Recommend 3 beers to the customer based on similarity and sentiment

In [45]:
top300_reviews_combined = top300_reviews.groupby('product_name')[['similarity','sentiment']].mean()
top300_reviews_combined['recommend'] = top300_reviews_combined['similarity']+top300_reviews_combined['sentiment']

In [46]:
top300_reviews_combined.sort_values(by='recommend', ascending=False)[0:3]

similarity  sentiment  recommend
product_name                                     
Emerald Grouper    0.764206     0.9961   1.760306
Flora              0.747172     0.9906   1.737772
Thicket            0.742931     0.9945   1.737431

### Recommend 3 beers to the customer based on similarity, sentiment and user rating

In [56]:
reviews_df['sentiment'] = reviews_df['product_review'].map(sentiment_analyzer_scores)

In [58]:
reviews_df_combined = reviews_df.groupby('product_name')[['similarity','sentiment', 'user_rating']].mean()
reviews_df_combined['recommend'] = reviews_df_combined['similarity']+reviews_df_combined['sentiment']

In [61]:
reviews_df_combined.sort_values(by = 'user_rating', ascending = False)[0:3]

similarity  sentiment  user_rating  recommend
product_name                                                              
Chemtrailmix                   0.591402   0.883917     4.816667   1.475318
Kentucky Brunch Brand Stout    0.568817   0.787348     4.812000   1.356165
It Was All A Dream             0.578521   0.600915     4.785000   1.179436

In [65]:
reviews_df_combined['recommend'].min()

1.1332525622299112

### Difference in the two recommendation methods
The recommendations change quite a bit between the two methods. One major reason is because we are using the entire dataset of reviews rather than those that are most similar to the attributes we are looking for. We are looking at all reviews for all products, regardless of the review's relevance to the input attributes (we used 'crisp', 'hoppy', and 'robust'). In general for something as subjective as beer preferences, using all of the reviews is not going to recommend anything personal to the user. Just because a lot of people like it doesn't mean that everyone likes it. For example, it's a hot topic right now whether or not pineapple should be a pizza topping. There are millions of people who love pineapple on pizza and will take to the internet to explain why it's great, but that doesn't mean that I will like pineapple pizza. I might prefer my pizza to be more savory and meaty rather than sweet, and that's exactly what is going on here. Just because lots of people like IPAs and that they are highly rated doesn't mean that everyone will like them- I've spat out a highly rated IPA before because I thought it was way to bitter and intense, which is why a lot of people liked it. This isn't about functionality- it's about taste, which differs person to person. That is why recommending the top rated beers without paying attention to similarity and sentiment isn't a good idea- you can see from the output above that the three top rated beers overall do not all fit well with our input attributes. One of them is actually one of the worst recommendation scores that we have in the dataset with a score of 1.179436. The other two perform closer to the mean recommendation of the set. Ultimately, blindly recommending high-rated beers will not meet the requirements of the user looking for recommendations. 